### PR TITLE
[ESP32] Option to configure the pool allocation from heap

### PR DESCRIPTION
**Pull request by shubhamdp**
There has been few occurrences where pool runs out of memory and we see `PacketBuffer: pool EMPTY`.

For the device which has external RAM, they can leverage the heap for allocating the buffer pool.

### CODE REVIEW

**Review by andy31415 — APPROVED** [review]

**Review by dhrishi — CHANGES_REQUESTED** [review]
Testing required as discussed

File: config/esp32/components/chip/Kconfig [hunk 1]
```diff
@@ -150,6 +150,14 @@ menu "CHIP Core"
             help
                 Option to enable/disable CHIP log level filtering APIs.
 
+        config CHIP_SYSTEM_CONFIG_POOL_USE_HEAP
+            bool "Use heap memory for system buffer pool"
+            default n
+            help
+                If enabled, all the allocations done using
+                chip::Platform::MemoryAlloc() will allocate memory from ESP32's heap area.
+                Device having the PSRAM can enable this.
```
Comment (bzbarsky-apple) [inline config/esp32/components/chip/Kconfig:159]: That is .... not what CHIP_SYSTEM_CONFIG_POOL_USE_HEAP does, though.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

**Comment by wqx6** [comment]
We should set `chip_config_memory_management` to `"platform"` and create platform `MemoryX()` functions to make the `chip::Platform::MemoryAlloc()` alloc memory on PSRAM.

**Comment by bzbarsky-apple** [comment]
There are two distinct changes under discussion here, it seems like:

1. Making Matter pools use heap on esp32. Note that this is not just a matter of where the memory lives; it also affects whether the max size of the pool is enforced.  When CONFIG_CHIP_SYSTEM_CONFIG_POOL_USE_HEAP is set, pools do _not_ enforce a max on the number of entries in the pool.
2. Changing Matter heap allocation on esp32 to allocate from a specific platform heap.

File: config/esp32/components/chip/Kconfig (hunk 1 above)
Comment (shubhamdp) [reply config/esp32/components/chip/Kconfig]: @bzbarsky-apple I have modified the help content, please take a look.

**Comment by shubhamdp** [comment]
> We should set `chip_config_memory_management` to `"platform"` and create platform `MemoryX()` functions to make the `chip::Platform::MemoryAlloc()` alloc memory on PSRAM.

All `MemoryAlloc()` for our platform are mapped to `malloc()` so new impl may not be required.

File: config/esp32/components/chip/Kconfig [hunk 2]
```diff
@@ -163,6 +163,13 @@ menu "CHIP Core"
             help
                 Option to enable/disable CHIP data model.
 
+        config CHIP_SYSTEM_CONFIG_POOL_USE_HEAP
+            bool "Use heap memory for system buffer pool"
+            default y
+            help
+                Option to use heap memory for system buffer pool.
```
Comment (bzbarsky-apple) [inline config/esp32/components/chip/Kconfig:170]: What is "system buffer pool"?

Setting CHIP_SYSTEM_CONFIG_POOL_USE_HEAP will make all ObjectPool instances:

1. Use heap.
2. Not actually enforce any size limits (is that what you want?)

It does not affect the behavior of SystemPacketBuffer, which is the first thing "system buffer" made me think of....

Comment (shubhamdp) [reply config/esp32/components/chip/Kconfig]: Sorry for the misunderstanding about what this option does. I have updated the prompt and help text.

This came up when we started seeing DRAM overflow in the all-clusters-app/esp32. On enabling this it adds about 19K of DRAM.

Setting this default to `n`. Can be used by whoever needs the DRAM but okay to allocate infinite number of object pools.

File: config/esp32/components/chip/Kconfig [hunk 3]
```diff
@@ -163,6 +163,16 @@ menu "CHIP Core"
...
 
+        config CHIP_SYSTEM_CONFIG_POOL_USE_HEAP
+            bool "Use heap memory to allocate object pools"
+            default n
+            help
+                This option enables the use of heap memory to allocate object pools.
+                When this option is enabled, it does not pre-allocate the object pools.
+                Additionally, there is no limit, (i.e. free heap is the limit) on the
+                maximum number of entries that can be allocated.
+                This option can be usefull if you run into static DRAM overflow.
```
Comment (bzbarsky-apple) [inline config/esp32/components/chip/Kconfig:174]: Note that:

1. This option can turn a failure to compile into just a device that does not properly implement the spec (and might fail cert, or worse yet pass cert and then fail to work with actual controllers).  This needs to be called out clearly.
2. The fact that there is no cap on pool sizes may lead to problems in places where embedded code currently assumes the pool size _is_ in fact limited and has no other limiters on the size of things.

Comment (shubhamdp) [reply config/esp32/components/chip/Kconfig]: I have added these points as a NOTE in the config options help.

Comment (shubhamdp) [reply config/esp32/components/chip/Kconfig]: @bzbarsky-apple can you please review the text again!

**Review by bzbarsky-apple — APPROVED** [review]

**Review by dhrishi — APPROVED** [review]

**Pull request merged**